syntax = "proto3";

option cc_enable_arenas = true;

package admin;

enum ObjectType {
  ASTEROID = 0;
  SHIP     = 1;
}

message Object {
  ObjectType type = 1;
  uint32 base_id  = 2;
  float  r        = 3;
  float  m        = 4;
  double x        = 5;
  double y        = 6;
  float  vx       = 7;
  float  vy       = 8;
}

message ObjectsList {
  repeated Object object = 1;
  uint32          left   = 2;
}


message Access {

  message Login {
    string login    = 1;
    string password = 2;
  }
  
  oneof choice {
    Login  login   = 1;
  
    uint64 success = 128;
    bool   fail    = 129;
  }
}


message SystemClock {

  enum Status {
    // Clock modes:
    MODE_REAL_TIME  = 0;
    MODE_DEBUG      = 1;
    MODE_TERMINATED = 2;
    
    // Clock statuses:
    CLOCK_IS_BUSY   = 3;
    INTERNAL_ERROR  = 4;
  }
  
  oneof choice {
    bool    time_req             = 1;
    bool    mode_req             = 2;
    bool    switch_to_real_time  = 3;
    bool    switch_to_debug_mode = 4;
    bool    terminate            = 5;
    uint32  tick_duration_us     = 6;
    uint32  proceed_ticks        = 7;
    
    uint64 now    = 129;
    Status status = 130;
  }
}



message Screen {

  enum Status {
    SUCCESS = 0;
    FAILED  = 1;
  }

  message Position {
    double x      = 2;
    double y      = 3;
    double width  = 4;
    double height = 5;
  }

  oneof choice {
    Position    move    = 1;
    ObjectType  show    = 2;
    
    Status      status  = 129;
    ObjectsList objects = 130;
  }
}


message Message {
  uint64 token = 1;

  oneof choice {
    // All possible interfaces are listed here
    Access      access       = 2;
    SystemClock system_clock = 3;
    Screen      screen       = 4;
  }
}
